libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -448,85 +895,23 @@ MQTTStatus_t MQTT_UpdateStatePublish( MQTTContext_t * pMqttContext,
     if( ( qos != MQTTQoS0 ) && ( mqttStatus == MQTTSuccess ) )
     {
         newState = MQTT_CalculateStatePublish( opType, qos );
-        isTransitionValid = validateTransitionPublish( currentState, newState, opType, qos );
-
-        if( isTransitionValid == true )
-        {
-            /* addRecord will check for collisions. */
-            if( opType == MQTT_RECEIVE )
-            {
-                mqttStatus = addRecord( pMqttContext->incomingPublishRecords,
-                                        MQTT_STATE_ARRAY_MAX_COUNT,
-                                        packetId,
-                                        qos,
-                                        newState );
-            }
-            /* Send operation. */
-            else
-            {
-                updateRecord( pMqttContext->outgoingPublishRecords, recordIndex, newState, false );
-            }
-        }
-        else
-        {
-            mqttStatus = MQTTIllegalState;
-            LogError( ( "Invalid transition from state %s to state %s.",
-                        MQTT_State_strerror( currentState ),
-                        MQTT_State_strerror( newState ) ) );
-        }
-
+        /* Validate state transition and update state records. */
+        mqttStatus = updateStatePublish( pMqttContext,
+                                         recordIndex,
+                                         packetId,
+                                         opType,
+                                         qos,
+                                         currentState,
+                                         &newState );
+
+        /* Output parameter. */
```

Comment:
Same as above, we could add a status check if we don't want to set it for errors